pkg/reconciler/apibinding/apibinding_reconcile_test.go

```diff
@@ -0,0 +1,745 @@
+					Status:   corev1.ConditionFalse,
+					Reason:   apisv1alpha1.WaitingForEstablishedReason,
+					Severity: conditionsv1alpha1.ConditionSeverityInfo,
+				},
+				{
+					Type:   apisv1alpha1.APIExportValid,
+					Status: corev1.ConditionTrue,
+				},
+			},
+		},
+	}
+
+	for testName, tc := range tests {
+		t.Run(testName, func(t *testing.T) {
+			createCRDCalled := false
+
+			r := &workspaceAPIExportReferenceReconciler{
+				getAPIExport: func(clusterName, name string) (*apisv1alpha1.APIExport, error) {
+					require.Equal(t, "org:some-workspace", clusterName)
+					require.Equal(t, "some-export", name)
+					return tc.apiExport, tc.getAPIExportError
+				},
+				getAPIResourceSchema: func(clusterName, name string) (*apisv1alpha1.APIResourceSchema, error) {
+					require.Equal(t, "org:some-workspace", clusterName)
+					return tc.apiResourceSchemas[name], tc.getAPIResourceSchemaError
+				},
+				getCRD: func(clusterName, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+					require.Equal(t, shadowWorkspaceName, clusterName)
+					return tc.crds[name], tc.getCRDError
+				},
+				createCRD: func(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, error) {
+					createCRDCalled = true
+					// We don't use the returned CRD - update in the future if we need it
+					return &apiextensionsv1.CustomResourceDefinition{}, tc.createCRDError
+				},
+				updateCRD: func(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, error) {
+					return nil, nil
+				},
+				deletedCRDTracker: &lockedStringSet{},
+			}
+
+			status, err := r.reconcile(context.Background(), tc.apiBinding)
+
+			gotErr := err != nil
+			require.Equal(t, tc.wantError, gotErr)
+			if tc.wantError {
+				return
+			}
+
+			require.Equal(t, status, tc.wantReconcileStatus)
+			require.Equal(t, tc.wantCreateCRD, createCRDCalled, "expected CRD to be created")
+
+			for _, wantCondition := range tc.wantConditions {
+				requireConditionMatches(t, tc.apiBinding, wantCondition)
+			}
+		})
+	}
+}
+
+// TODO(ncdc): this is a modified copy from apibinding admission. Unify these into a reusable package.
+type bindingBuilder struct {
+	apisv1alpha1.APIBinding
+}
+
+func (b *bindingBuilder) DeepCopy() *bindingBuilder {
+	return &bindingBuilder{
+		APIBinding: *b.APIBinding.DeepCopy(),
+	}
+}
+
+func (b *bindingBuilder) Build() *apisv1alpha1.APIBinding {
+	return b.APIBinding.DeepCopy()
+}
+
+func (b *bindingBuilder) WithClusterName(clusterName string) *bindingBuilder {
+	b.ClusterName = clusterName
+	return b
+}
+
+func (b *bindingBuilder) WithName(name string) *bindingBuilder {
+	b.Name = name
+	return b
+}
+
+func (b *bindingBuilder) WithoutWorkspaceReference() *bindingBuilder {
+	b.Spec.Reference.Workspace = nil
+	return b
+}
+
+func (b *bindingBuilder) WithWorkspaceReference(workspaceName, exportName string) *bindingBuilder {
```

Comment:
(No action required) I think part of the advantage of builders is removing the need to parametrize them where possible. Ideally the cases would be as simple as `WithSomeWorkspaceReference` and `WithAnotherWorkspaceReference` with correspondence across builders (i.e. SomeWorkspace and AnotherWorkspace constants). This minimizes the details in the test cases to the minimum (Some, Another, Different, etc) required to clearly indicate intent.